pgjdbc/src/main/java/org/postgresql/hostchooser/MultiHostChooser.java

```diff
@@ -56,32 +56,40 @@ public Iterator<CandidateHost> iterator() {
   }
 
   private Iterator<CandidateHost> candidateIterator() {
-    if (targetServerType != HostRequirement.preferSecondary) {
+    if (   targetServerType != HostRequirement.preferSecondary
+        && targetServerType != HostRequirement.preferPrimary   ) {
       return getCandidateHosts(targetServerType).iterator();
     }
 
+    HostRequirement preferredServerType =
+        targetServerType == HostRequirement.preferSecondary
+          ? HostRequirement.secondary
+          : HostRequirement.primary;
+
     // preferSecondary tries to find secondary hosts first
     // Note: sort does not work here since there are "unknown" hosts,
     // and that "unknown" might turn out to be master, so we should discard that
     // if other secondaries exist
-    List<CandidateHost> secondaries = getCandidateHosts(HostRequirement.secondary);
+    // Same logic as the above works for preferPrimary if we replace "secondary"
+    // with "primary" and vice versa
+    List<CandidateHost> preferred = getCandidateHosts(preferredServerType);
     List<CandidateHost> any = getCandidateHosts(HostRequirement.any);
 
-    if (secondaries.isEmpty()) {
+    if (preferred.isEmpty()) {
       return any.iterator();
     }
 
     if (any.isEmpty()) {
-      return secondaries.iterator();
+      return preferred.iterator();
     }
 
```

Comment:
Can you explain the logic here ?

if preferred is empty fails and any is empty is true, then it's going to return an empty iterator ?

Reply:
If `preferred` is not empty, then neither can `any` be.

I'm guessing a static analysis would flag the `any.get(0)` without checking as a bug.
An `assert !any.isEmpty()` should suffice.

Reply:
> Can you explain the logic here ?
> 
> if preferred is empty fails and any is empty is true, then it's going to return an empty iterator ?

if preferred-is-empty fails (i.e. "preferred" iterator is not empty) and any-is-empty is true (i.e. "any" iterator is empty), then it's going to return "preferred" iterator, which is not empty. It follows exactly same logic for the "preferred" iterator that was here before for the "secondaries" iterator.

I agree with @OrangeDog though that if "preferred" is not empty, then "any" cannot be empty either, which means that "any-is-empty is true" branch will never be entered

Reply:
None of the checks are actually necessary:

```
List<CandidateHost> preferred = getCandidateHosts(preferredServerType);
List<CandidateHost> any = getCandidateHosts(HostRequirement.any);

any.removeAll(preferred);
return append(preferred, any).iterator();
```

May need a different function call if the lists aren't mutable. Or skip `removeAll` and don't worry about repeated attempts.

Reply:
> Either way this code seems difficult to follow. Can someone suggest a simplification?

Pushed [404efd1](https://github.com/pgjdbc/pgjdbc/pull/2483/commits/404efd1029e3eb92bc3b759585793254aca6764e) commit that simplifies `MultiHostChooser.candidateIterator()` code and fixes condition for the optimization inside this method